File: worker/encryption.py
"""Encryption utilities for the worker."""

import base64
import hashlib
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.scrypt import Scrypt
from cryptography.hazmat.backends import default_backend
import os

from config import get_settings

settings = get_settings()

SALT_LENGTH = 32
IV_LENGTH = 16
AUTH_TAG_LENGTH = 16


def _get_master_key() -> bytes:
    """Get the master encryption key."""
    return base64.b64decode(settings.encryption_master_key)


def _derive_key(master_key: bytes, salt: bytes) -> bytes:
    """Derive a key from master key and salt using scrypt."""
    kdf = Scrypt(
        salt=salt,
        length=32,
        n=2**14,
        r=8,
        p=1,
        backend=default_backend()
    )
    return kdf.derive(master_key)


def encrypt(plaintext: str) -> str:
    """Encrypt a string using AES-256-GCM."""
    master_key = _get_master_key()
    salt = os.urandom(SALT_LENGTH)
    key = _derive_key(master_key, salt)
    iv = os.urandom(IV_LENGTH)
    
    aesgcm = AESGCM(key)
    ciphertext = aesgcm.encrypt(iv, plaintext.encode('utf-8'), None)
    
    # Combine: salt + iv + ciphertext (includes auth tag)
    combined = salt + iv + ciphertext
    return base64.b64encode(combined).decode('utf-8')


def decrypt(encrypted_data: str) -> str:
    """Decrypt an encrypted string."""
    master_key = _get_master_key()
    combined = base64.b64decode(encrypted_data)
    
    # Extract components
    salt = combined[:SALT_LENGTH]
    iv = combined[SALT_LENGTH:SALT_LENGTH + IV_LENGTH]
    ciphertext = combined[SALT_LENGTH + IV_LENGTH:]
    
    key = _derive_key(master_key, salt)
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, ciphertext, None)
    
    return plaintext.decode('utf-8')


def hash_content(content: str) -> str:
    """Hash content for deduplication."""
    return hashlib.sha256(content.encode('utf-8')).hexdigest()

File: src/lib/encryption.ts
import { createCipheriv, createDecipheriv, randomBytes, scryptSync } from "crypto";

const ALGORITHM = "aes-256-gcm";
const IV_LENGTH = 16;
const AUTH_TAG_LENGTH = 16;
const SALT_LENGTH = 32;

function getMasterKey(): Buffer {
  const key = process.env.ENCRYPTION_MASTER_KEY;
  if (!key) {
    throw new Error("ENCRYPTION_MASTER_KEY environment variable is not set");
  }
  return Buffer.from(key, "base64");
}

function deriveKey(masterKey: Buffer, salt: Buffer): Buffer {
  return scryptSync(masterKey, salt, 32);
}

/**
 * Encrypts a string using AES-256-GCM with a derived key
 * Format: base64(salt + iv + authTag + ciphertext)
 */
export function encrypt(plaintext: string): string {
  const masterKey = getMasterKey();
  const salt = randomBytes(SALT_LENGTH);
  const key = deriveKey(masterKey, salt);
  const iv = randomBytes(IV_LENGTH);

  const cipher = createCipheriv(ALGORITHM, key, iv);
  const encrypted = Buffer.concat([
    cipher.update(plaintext, "utf8"),
    cipher.final(),
  ]);
  const authTag = cipher.getAuthTag();

  // Combine: salt + iv + authTag + ciphertext
  const combined = Buffer.concat([salt, iv, authTag, encrypted]);
  return combined.toString("base64");
}

/**
 * Decrypts an encrypted string
 */
export function decrypt(encryptedData: string): string {
  const masterKey = getMasterKey();
  const combined = Buffer.from(encryptedData, "base64");

  // Extract components
  const salt = combined.subarray(0, SALT_LENGTH);
  const iv = combined.subarray(SALT_LENGTH, SALT_LENGTH + IV_LENGTH);
  const authTag = combined.subarray(
    SALT_LENGTH + IV_LENGTH,
    SALT_LENGTH + IV_LENGTH + AUTH_TAG_LENGTH
  );
  const ciphertext = combined.subarray(SALT_LENGTH + IV_LENGTH + AUTH_TAG_LENGTH);

  const key = deriveKey(masterKey, salt);
  const decipher = createDecipheriv(ALGORITHM, key, iv);
  decipher.setAuthTag(authTag);

  const decrypted = Buffer.concat([
    decipher.update(ciphertext),
    decipher.final(),
  ]);

  return decrypted.toString("utf8");
}

/**
 * Encrypts sensitive data, returns null if input is null/undefined
 */
export function encryptIfPresent(value: string | null | undefined): string | null {
  if (!value) return null;
  return encrypt(value);
}

/**
 * Decrypts sensitive data, returns null if input is null/undefined
 */
export function decryptIfPresent(value: string | null | undefined): string | null {
  if (!value) return null;
  try {
    return decrypt(value);
  } catch {
    console.error("Failed to decrypt value");
    return null;
  }
}

/**
 * Hashes content for deduplication (not for security)
 */
export function hashContent(content: string): string {
  const { createHash } = require("crypto");
  return createHash("sha256").update(content).digest("hex");
}
